From a0a194b3e83873bc5c348e99d610e4485b824f53 Mon Sep 17 00:00:00 2001
From: "Miss Islington (bot)"
 <31488909+miss-islington@users.noreply.github.com>
Date: Fri, 9 Aug 2024 18:55:36 +0200
Subject: [PATCH 17/24] [3.12] gh-79846: Make ssl.create_default_context()
 ignore invalid certificates (GH-91740) (#122769)
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

An error in one certificate should not cause the whole thing to fail.

(cherry picked from commit 9e551f9b351440ebae79e07a02d0e4a1b61d139e)

Co-authored-by: pukkandan <pukkandan.ytdlp@gmail.com>
Co-authored-by: Serhiy Storchaka <storchaka@gmail.com>
Signed-off-by: Michał Górny <mgorny@gentoo.org>
---
 Lib/ssl.py | 9 ++++-----
 1 file changed, 4 insertions(+), 5 deletions(-)

diff --git a/Lib/ssl.py b/Lib/ssl.py
index 7825cccba9b..67688299b91 100644
--- a/Lib/ssl.py
+++ b/Lib/ssl.py
@@ -575,18 +575,17 @@ def set_alpn_protocols(self, alpn_protocols):
         self._set_alpn_protocols(protos)
 
     def _load_windows_store_certs(self, storename, purpose):
-        certs = bytearray()
         try:
             for cert, encoding, trust in enum_certificates(storename):
                 # CA certs are never PKCS#7 encoded
                 if encoding == "x509_asn":
                     if trust is True or purpose.oid in trust:
-                        certs.extend(cert)
+                        try:
+                            self.load_verify_locations(cadata=cert)
+                        except SSLError as exc:
+                            warnings.warn(f"Bad certificate in Windows certificate store: {exc!s}")
         except PermissionError:
             warnings.warn("unable to enumerate Windows certificate store")
-        if certs:
-            self.load_verify_locations(cadata=certs)
-        return certs
 
     def load_default_certs(self, purpose=Purpose.SERVER_AUTH):
         if not isinstance(purpose, _ASN1Object):
